src/contract_client.rs

```diff
@@ -864,6 +1159,72 @@ impl<Type> ContractClient<Type> {
         }
     }
 
+    /// Like [`dry_run_update_with_reject_reason_info`](Self::dry_run_update_with_reject_reason_info) but expects an already
+    /// formed parameter.
+    pub async fn dry_run_update_raw_with_reject_reason_info<E>(
```

Comment:
When seeing the consumer side of this (https://github.com/Concordium/concordium-dapp-examples/pull/87#discussion_r1671672842), I don't know if I think this is the right approach.

It feels like `InvokeContractOutcome` is a pseudo-result but without the ergonomics that come with just using a plain`Result`. I wonder if it's just better to return both `RejectReason` and an added DecodedRejectReason? Or is this too weird?

Alternatively we could rely on users decoding the `RejectReason` manually (maybe make the function easier to discover by adding it the `impl RejectReason`?), but this might be too implicit?

Reply:
One complication is that we need the `returnValue` to decode the smart contract error with the schema reliably. This was not exposed previously.

Reply:
What are your thoughts about the first suggestion here then with regards to adding `+ From<DecodedRejectReason>` to the list required for `E` to implement? The downside here is that the list of types to implement `From<...>` on the consuming end gets a bit long.. We have a convenience error type `ViewError` for this purpose for the `dry_run_update` functions, but updating it would obviously break things and adding yet another convenience error type might be a path we want to avoid...?

Reply:
> One complication is that we need the `returnValue` to decode the smart contract error with the schema reliably (contracts might or might not derive `Reject` on the errorEnum). This was not exposed previously and is not exposed in the `RejectReason` type. Otherwise, I like the idea that the `RejectReason` can be decoded directly.

I also feel like it would be misplaced in `impl RejectReason` for this reason

Reply:
Off-chain discussions with Emil G and Soeren with great input about the historical reasoning:
```
/// `Return_values` vs. `error_codes` in rejected transactions:
/// Historically, Concordium had smart contracts V0 which had no `retun_value`.
/// Error codes (negative values used commonly in computer science to represent
/// errors) were chosen as a method to report the reject reason. Smart
/// contracts could only revert using one "[Unspecified (Default reject)]" error
/// and the error code (i32::MIN) was used to represent this error. This
/// "[Unspecified (Default reject)]" error definition still exists in the
/// `concordium-std` crate.
///
/// `Return_values` were introduced in smart contract V1 and smart contracts
/// were fitted with features to revert with different reasons (as defined in
/// the smart contract logic). The `return_value` is used to distinguish between
/// the different reasons for the revert coming from the smart contract logic.
///
/// There are some older types used by the node (such as `RejectReason`) that
/// only include the `error_code` but not the `return_value`. The reason can be
/// explained by the historical development of smart contracts in Concordium to
/// avoid breaking changes as well as to keep the size of the nodes as small as
/// possible since the `return_value` does not need to be saved by the nodes for
/// achieving consensus.
///
/// As a result, this decoding function needs both the `return_value` and the
/// `error_code` to decode the `reject_reason` into a human-readable error.
```

Reply:
@soerenbf 
I got rid of the `+ From<DecodedRejectReason>` conversion altogether.

Reply:
@soerenbf Regarding your question why the `contract_client` uses functions with error generics (e.g.  `E: From<...> ... `) instead of hardcoding the error to `ViewError`. My explanation would be that the variants in `ViewError` are likely already implemented by the error used on the consuming side of the SDK as such the consuming side could use their error type without implementing additional conversions.
https://github.com/Concordium/concordium-rust-sdk/blob/32eb030553efa4c9be3cb7a3831095cd57c9d67e/src/contract_client.rs#L93